Add Harmony Hub Extension - Control Logitech Harmony Devices

chadrwalters opened the pull request:
## Description

This extension enables Raycast users to control their Logitech Harmony Hub devices directly from Raycast. Key features include:

* 🔍 Automatic hub discovery on your local network
* 📱 Control all your Harmony-connected devices
* ⚡️ Quick access to device commands
* 🎮 Start and stop activities
* 🔄 Real-time status updates
* ⌨️ Full keyboard navigation

The extension provides a seamless experience for Harmony Hub users, allowing them to control their entertainment devices without leaving their keyboard.

## Screenshots

Here are the key screens of the extension in action:

![Automatic discovery of Harmony Hubs on your local network, making setup effortless]
![harmony-control-1](https://github.com/user-attachments/assets/9573719f-a82c-410a-9aba-149c91bfafb0)

![Choose and connect to your preferred Harmony Hub when multiple hubs are available]
![harmony-control-2](https://github.com/user-attachments/assets/01274252-68c1-4221-8ed8-c309270749d0)

![Browse and select from all your Harmony-configured devices in an intuitive interface]![harmony-control-3](https://github.com/user-attachments/assets/a0c09877-5159-4026-b5b0-9e18def17212)

![Access device-specific commands quickly with a keyboard-optimized command list]
![harmony-control-4](https://github.com/user-attachments/assets/4e8893da-50e6-4b3f-840b-e8de5e277a53)


## Checklist

* [x] I read the extension guidelines
* [x] I read the documentation about publishing
* [x] I ran `npm run build` and tested this distribution build in Raycast
* [x] I checked that files in the `assets` folder are used by the extension itself
* [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer reviewed (commented):
I guess this is made with AI too. Which editor are you using?

chadrwalters commented:
> I guess this is made with AI too. Which editor are you using?

This one I did with windsurf

chadrwalters commented:
# PR Updates: Code Quality and Error Handling Enhancements

I've completed all requested changes from the PR feedback. Here's a comprehensive summary of the improvements:

## 🎨 Code Quality Improvements

1. **Variable Naming (`useActivityFiltering.ts`)**
   - Renamed shadowed `activities` variable to `typeActivities` for better clarity
   - Improved readability in the `activitiesByType` useMemo hook

2. **State Management (`HarmonyCommand.tsx`)**
   - Removed unnecessary `setTimeout(fn, 0)` calls
   - Implemented direct state updates for better performance
   - Improved state transition handling

3. **Conditional Rendering (`HubsView.tsx`)**
   - Removed redundant conditional checks for `refresh` and `clearCache` actions
   - Simplified component logic
   - Added `isViewLoading` helper function for cleaner loading state management

## 🛡️ Error Handling Improvements

1. **Centralized Toast Management**
   - Created `ToastManager` service for consistent toast handling
   - Integrated `showFailureToast` from `@raycast/utils`
   - Added support for success, error, warning, and loading toasts
   - Implemented toast logging configuration

2. **Error Handler Improvements**
   - Enhanced `ErrorHandler` service with comprehensive error categorization
   - Added support for async operations with proper error handling
   - Implemented recovery actions based on error categories
   - Added context-aware error messages

3. **Harmony Manager Updates**
   - Improved hub verification process
   - Enhanced error handling in hub discovery
   - Added proper cleanup of resources
   - Implemented robust cache management

## 🧪 Testing Notes
- All changes have been tested locally
- Error handling has been verified with various error scenarios
- Toast notifications are working consistently
- State management improvements have been validated

## 📝 Additional Notes
- Maintained TypeScript type safety throughout changes
- Added comprehensive documentation for new functionality
- Ensured backward compatibility
- Followed Raycast's best practices for error handling and user feedback

Please let me know if you'd like me to clarify or expand on any of these changes.

pernielsentikaer commented on extensions/harmonyhub/.eslintrc.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": "@raycast/eslint-config",
+  "rules": {
+    "@typescript-eslint/no-explicit-any": "warn",
+    "@typescript-eslint/explicit-function-return-type": "off"
+  }
+} 
```
```suggestion
{
  "root": true,
  "extends": ["@raycast"]
}

```

pernielsentikaer commented on extensions/harmonyhub/.prettierrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "printWidth": 120,
+  "singleQuote": false,
+  "trailingComma": "es5",
+  "tabWidth": 2
+} 
```
```suggestion
{
  "printWidth": 120,
  "singleQuote": false
}

```

pernielsentikaer commented on extensions/harmonyhub/CHANGELOG.md:
```diff
@@ -0,0 +1,34 @@
+# Changelog
```
```suggestion
# Harmony Changelog
```

pernielsentikaer reviewed (changes requested):
Do you know how to code, or are you completely reliant on the AI editor to do it for you? It seems like much of it could be done much more simply. Reviewing extensions made with AI takes a lot of time

@greptilai check again

chadrwalters commented:
> Do you know how to code, or are you completely reliant on the AI editor to do it for you? It seems like much of it could be done much more simply. Reviewing extensions made with AI takes a lot of time
> 
> @greptilai check again

@pernielsentikaer older dev here. Been a useless VP for 15 years, without code in prod for 10 years until the last 6 months. Doing a lot more with AI now and on the forefront for helping my teams start to get use from it.

Last thing I am trying to do is waste your time/take you away from more important things. If this is something you would like me to stop, I will. I created the rules for the AI based on the SDK and I am starting to improve them based on the PR comments I am getting. But I get it man, I love your product and have it so integrated in my life to me it was an easy outlet for me to test out tools and techniques for "real".

pernielsentikaer commented:
Could you fix the CI errors ;)

chadrwalters commented on extensions/harmonyhub/src/services/toast.ts:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    await showFailureToast({ title, message });
+  }
```
moved to a animated toast away from the failure.

chadrwalters commented:
Thank you for the detailed review. I've addressed all the feedback points with the following improvements:

1. Code Quality & Clarity:
   - Added "Harmony Control" subtitle to the command in package.json for better discoverability
   - Removed unused `renderItem` prop from ActivitiesView to reduce confusion and improve code clarity
   - Improved error handler to properly preserve stack traces for better debugging

2. Toast Notifications:
   - Modified warning() method to use Toast.Style.Animated instead of showFailureToast to better differentiate warnings from errors
   - Maintained use of @raycast/utils showFailureToast for error handling as it aligns with Raycast's best practices
   - Implemented clear visual hierarchy:
     * Errors: Red (showFailureToast)
     * Warnings: Neutral gray (Toast.Style.Animated)
     * Success: Green (Toast.Style.Success)

3. State Management:
   - Updated HubsView cancel handler to use refresh instead of clearCache for more predictable behavior
   - Improved progress toast implementation to handle percentage display more elegantly

4. Documentation:
   - Enhanced JSDoc comments for better code understanding
   - Updated method descriptions to accurately reflect their behavior

pernielsentikaer commented:
Even if all your replies are made with AI, it feels like most of the time it actually makes it worse
